troposphere/haze/source/ptp_responder.cpp

```diff
@@ -0,0 +1,698 @@
+
+                    storage_info.max_capacity = total_space;
+                    storage_info.free_space_in_bytes = free_space;
+                    storage_info.free_space_in_images = 0;
+                    storage_info.storage_description = "SD Card";
+                }
+                break;
+
+            default:
+                R_THROW(haze::ResultStorageNotFound());
+        }
+
+        /* Write the result. */
+        R_TRY(db.WriteVariableLengthData(m_request_header, [&] {
+            R_TRY(db.Add(storage_info.storage_type));
+            R_TRY(db.Add(storage_info.filesystem_type));
+            R_TRY(db.Add(storage_info.access_capability));
+            R_TRY(db.Add(storage_info.max_capacity));
+            R_TRY(db.Add(storage_info.free_space_in_bytes));
+            R_TRY(db.Add(storage_info.free_space_in_images));
+            R_TRY(db.AddString(storage_info.storage_description));
+            R_TRY(db.AddString(storage_info.volume_label));
+
+            R_SUCCEED();
+        }));
+
+        R_RETURN(this->WriteResponse(PtpResponseCode_Ok));
+    }
+
+    Result PtpResponder::GetObjectHandles(PtpDataParser &dp) {
+        PtpDataBuilder db(s_bulk_write_buffer, std::addressof(m_usb_server));
+
+        /* Get the object ID the client requested enumeration for. */
+        u32 storage_id, object_format_code, association_object_handle;
+        R_TRY(dp.Read(std::addressof(storage_id)));
+        R_TRY(dp.Read(std::addressof(object_format_code)));
+        R_TRY(dp.Read(std::addressof(association_object_handle)));
+        R_TRY(dp.Finalize());
+
+        /* Handle top-level requests. */
+        if (storage_id == PtpGetObjectHandles_AllStorage) {
+            storage_id = StorageId_SdmcFs;
+        }
+
+        /* Rewrite requests for enumerating storage directories. */
+        if (association_object_handle == PtpGetObjectHandles_RootParent) {
+            association_object_handle = storage_id;
+        }
+
+        /* Check if we know about the object. If we don't, it's an error. */
+        auto *fileobj = m_object_database.GetObject(association_object_handle);
+        R_UNLESS(fileobj != nullptr, haze::ResultObjectNotFound());
+
+        /* Try to read the object as a directory. */
+        FsDir dir;
+        R_TRY(m_fs.OpenDirectory(fileobj->GetName(), FsDirOpenMode_ReadDirs | FsDirOpenMode_ReadFiles, std::addressof(dir)));
+
+        /* Ensure we maintain a clean state on exit. */
+        ON_SCOPE_EXIT { m_fs.DirectoryClose(std::addressof(dir)); };
+
+        /* Count how many entries are in the directory. */
+        s64 entry_count = 0;
+        R_TRY(m_fs.DirectoryGetEntryCount(std::addressof(dir), std::addressof(entry_count)));
+
+        /* Begin writing. */
+        R_TRY(db.AddDataHeader(m_request_header, sizeof(u32) + (entry_count * sizeof(u32))));
+        R_TRY(db.Add(static_cast<u32>(entry_count)));
+
+        /* TODO: How should we handle the directory contents changing during enumeration? */
```

Comment:
In addition to the question-remark comment, this probably wants a behavioral/action comment.